### PR TITLE
[core] Make ClasspathClassLoader::getResource child first

### PR DESCRIPTION
**Pull request by oowekyala**
## Describe the PR

This makes it so that one can use the auxclasspath to include a standard lib for a different java version than the one on PMD's own classpath.

## Related issues

- Related to #4620

## Ready?

- [ ] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

File: pmd-java/src/test/java/net/sourceforge/pmd/lang/java/symbols/ClassLoadingChildFirstTest.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+     * The test asserts that when putting that jar on the auxclasspath,
+     * a {@link TypeSystem} prefers that custom class to the one on the
+     * bootstrap classpath. The functionality under test is actually in
+     * {@link ClasspathClassLoader}.
+     */
+    @Test
+    void testClassLoading() {
+        String jarPath = getClass().getPackage().getName().replace('.', '/') + "/custom_java_lang.jar";
+        URL url = Thread.currentThread().getContextClassLoader().getResource(jarPath);
+        Path file = Paths.get(url.getPath());
```
Comment (adangel) [inline pmd-java/src/test/java/net/sourceforge/pmd/lang/java/symbols/ClassLoadingChildFirstTest.java:41]: This look a bit complicated. Wouldn't the following work?

```java
Path file = Paths.get("src/test/resources", getClass().getPackage().getName().replace('.', '/'), "custom_java_lang.jar");
```

File: pmd-java/src/test/java/net/sourceforge/pmd/lang/java/symbols/ClassLoadingChildFirstTest.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+        PMDConfiguration config = new PMDConfiguration();
+        config.prependAuxClasspath(file.toAbsolutePath().toString());
+
+        TypeSystem typeSystem = TypeSystem.usingClassLoaderClasspath(config.getClassLoader());
+
+        JClassType voidClass = typeSystem.BOXED_VOID;
+        List<JMethodSymbol> declaredMethods = voidClass.getSymbol().getDeclaredMethods();
+        assertThat(declaredMethods, hasSize(1));
+        assertThat(declaredMethods.get(0), hasProperty("simpleName", equalTo("customMethodOnJavaLangVoid")));
```
Comment (adangel) [inline pmd-java/src/test/java/net/sourceforge/pmd/lang/java/symbols/ClassLoadingChildFirstTest.java:51]: Maybe we should at least provide the sources of custom_java_lang.jar here as a comment. If I understand it correctly, it looks like this:

```java
package java.lang;
public final class Void {
    public static final Class<Void> TYPE = (Class<Void>) Class.getPrimitiveClass("void");
    private Void() {}
    public void customMethodOnJavaLangVoid() {}
}
```

Maybe we could simply add the source additionally into the jar?

**Review by adangel — APPROVED** [review]
Thanks!

I'll look at the failing Windows unit test (`ClassLoadingChildFirstTest.testClassLoading:41 � InvalidPath Illegal char <:> at index 2: /D:/a/pmd/pmd/pmd-java/target/test-classes/net/sourceforge/pmd/lang/java/symbols/custom_java_lang.jar`) and merge it afterwards.

File: pmd-java/src/test/java/net/sourceforge/pmd/lang/java/symbols/ClassLoadingChildFirstTest.java (hunk 2 above)
Comment (adangel) [reply pmd-java/src/test/java/net/sourceforge/pmd/lang/java/symbols/ClassLoadingChildFirstTest.java:51]: I see, you already added the sources to the jar. Perfect!

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 0 violations,<br />
introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 2 violations, 0 errors and 0 configuration errors.<br />
<a href="Chunk.io now requires authentication. Send an email to hello@chunk.io to get an invite./diff1/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Pull request merged**